src/libs/Network.js

```diff
@@ -26,80 +28,84 @@ const [onResponse, registerResponseHandler] = createCallback();
 const [onError, registerErrorHandler] = createCallback();
 const [onRequestSkipped, registerRequestSkippedHandler] = createCallback();
 
-let didLoadPersistedRequests;
-let isOffline;
+function processRequest(request) {
+    const finalParameters = _.isFunction(enhanceParameters)
+        ? enhanceParameters(request.command, request.data)
+        : request.data;
 
-const PROCESS_REQUEST_DELAY_MS = 1000;
+    onRequest(request, finalParameters);
+    return HttpUtils.xhr(request.command, finalParameters, request.type, request.shouldUseSecure);
+}
 
-/**
- * Process the offline NETWORK_REQUEST_QUEUE
- * @param {Array<Object> | null} persistedRequests - Requests
- */
-function processOfflineQueue(persistedRequests) {
-    // NETWORK_REQUEST_QUEUE is shared across clients, thus every client will have similiar copy of
-    // NETWORK_REQUEST_QUEUE. It is very important to only process the queue from leader client
-    // otherwise requests will be duplicated.
-    // We only process the persisted requests when
-    // a) Client is leader.
-    // b) User is online.
-    // c) requests are not already loaded,
-    // d) When there is at least one request
-    if (!ActiveClientManager.isClientTheLeader()
-        || isOffline
-        || didLoadPersistedRequests
-        || !persistedRequests
-        || !persistedRequests.length) {
+function processPersistedRequestsQueue() {
+    const persistedRequests = NetworkRequestQueue.getPersistedRequests();
+
+    // This sanity check is also a recursion exit point
+    if (isOffline || _.size(persistedRequests) === 0) {
+        return Promise.resolve();
+    }
+
+    const tasks = _.map(persistedRequests, request => processRequest(request)
+        .then((response) => {
+            if (response.jsonCode !== CONST.NETWORK.SUCCESS_CODE) {
+                throw new Error('Persisted request failed');
+            }
+
+            NetworkRequestQueue.removeRetryableRequest(request);
+        })
+        .catch(() => {
+            const retryCount = NetworkRequestQueue.incrementRetries(request);
+            if (retryCount >= CONST.NETWORK.MAX_PERSISTED_REQUEST_RETRIES) {
+                // Request failed too many times removing from persisted storage
+                NetworkRequestQueue.removeRetryableRequest(request);
+            }
+        }));
+
+    // Do a recursive call in case the queue is not empty after processing the current batch
+    return Promise.all(tasks)
+        .then(processPersistedRequestsQueue);
+}
```

Comment:
We can have remaining requests when 
- the request failed for some reason and is not removed from the queue - as seen in the catch block
- we became offline again, which would again lead to the catch block above. In that case the queue would still stop and will be retriggered once we're back online, more requests can be added to it in the meantime

Not sure if we should capture that in a comment or something